tests/xharness/TestReporter.cs

```diff
@@ -0,0 +1,490 @@
+			doc.Load (reader);
+			foreach (XmlNode node in doc.SelectNodes ($"/root/processes/item[pid = '" + pid + "']")) {
+				Console.WriteLine (node?.InnerXml);
+				Console.WriteLine (node?.SelectSingleNode ("reason")?.InnerText);
+				crashReason = node?.SelectSingleNode ("reason")?.InnerText;
+			}
+		}
+
+		// return if the tcp connection with the device failed
+		async Task<bool> TcpConnectionFailed ()
+		{
+			using var reader = new StreamReader (runner.MainLog.FullPath);
+			string line;
+			while ((line = await reader.ReadLineAsync ()) != null) {
+				if (line.Contains ("Couldn't establish a TCP connection with any of the hostnames")) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		// kill any process 
+		Task KillAppProcess (int pid, CancellationTokenSource cancellationSource) { 
+				var launchTimedout = cancellationSource.IsCancellationRequested;
+				var timeoutType = launchTimedout ? "Launch" : "Completion";
+				var timeoutValue = launchTimedout ? runner.LaunchTimeout : Timeout.TotalSeconds;
+
+				runner.MainLog.WriteLine ($"{timeoutType} timed out after {timeoutValue} seconds");
+				return ProcessManager.KillTreeAsync (pid, runner.MainLog, true);
+		}
+
+		async Task CollectResult (Task<ProcessExecutionResult> processExecution)
+		{ 
+			// wait for the execution of the process, once that is done, perform all the parsing operations and
+			// leave a clean API to be used by AppRunner, hidding all the diff details
+			var result = await processExecution;
+			if (!waitedForExit && !result.TimedOut) {
+				// mlaunch couldn't wait for exit for some reason. Let's assume the app exits when the test listener completes.
+				runner.MainLog.WriteLine ("Waiting for listener to complete, since mlaunch won't tell.");
+				if (!await listener.CompletionTask.TimeoutAfter (Timeout - TimeoutWatch.Elapsed)) {
+					result.TimedOut = true;
+				}
+			}
+
+			if (result.TimedOut) {
+				timedout = true;
+				Success = false;
+				runner.MainLog.WriteLine (timeoutMessage, Timeout.TotalMinutes);
+			} else if (result.Succeeded) {
+				runner.MainLog.WriteLine (completionMessage);
+				Success = true;
+			} else {
+				runner.MainLog.WriteLine (failureMessage);
+				Success = false;
+			}
+		}
+
+		public void LaunchCallback (Task<bool> launchResult)
+		{
+			if (launchResult.IsFaulted) {
+				runner.MainLog.WriteLine ("Test launch failed: {0}", launchResult.Exception);
+			} else if (launchResult.IsCanceled) {
+				runner.MainLog.WriteLine ("Test launch was cancelled.");
+			} else if (launchResult.Result) {
+				runner.MainLog.WriteLine ("Test run started");
+			} else {
+				cancellationTokenSource.Cancel ();
+				runner.MainLog.WriteLine ("Test launch timed out after {0} minute(s).", runner.LaunchTimeout);
+				timedout = true;
+			}
+		}
+
+		public async Task CollectSimulatorResult (Task<ProcessExecutionResult> processExecution)
+		{
+			isSimulatorTest = true;
+			await CollectResult (processExecution);
+
+			if (!Success.Value) {
```

Comment:
The code is very similar between Sim and Device, we had code duplication that has been moved to CollectResult, only diff between sim and device is killing the process.